src/adapters/MorphoMarketV1Adapter.sol

```diff
@@ -94,54 +107,49 @@ contract MorphoMarketV1Adapter is IMorphoMarketV1Adapter {
+        require(data.length == 0, InvalidData());
         require(msg.sender == parentVault, NotAuthorized());
-        require(marketParams.loanToken == asset, LoanAssetMismatch());
 
-        if (assets > 0) IMorpho(morpho).withdraw(marketParams, assets, 0, address(this), address(this));
+        uint256 shares;
+        if (assets > 0) {
+            (, shares) = IMorpho(morpho).withdraw(marketParams(), assets, 0, address(this), address(this));
+            // Safe cast because Market V1 bounds the total shares to uint128.max.
+            supplyShares -= uint128(shares);
+        }
 
-        uint256 oldAllocation = allocation(marketParams);
-        uint256 newAllocation = MorphoBalancesLib.expectedSupplyAssets(IMorpho(morpho), marketParams, address(this));
-        updateList(marketParams, oldAllocation, newAllocation);
+        uint256 _newAllocation = newAllocation();
+        // Safe casts because Market V1 bounds totalSupplyAssets to uint128.max.
+        int256 change = int256(_newAllocation) - int256(uint256(allocation));
+        allocation = uint128(_newAllocation);
 
-        // Safe casts because Market V1 bounds the total supply of the underlying token, and allocation is less than the
-        // max total assets of the vault.
-        return (ids(marketParams), int256(newAllocation) - int256(oldAllocation));
-    }
+        emit Deallocate(shares);
 
-    function updateList(MarketParams memory marketParams, uint256 oldAllocation, uint256 newAllocation) internal {
-        if (oldAllocation > 0 && newAllocation == 0) {
-            Id marketId = marketParams.id();
-            for (uint256 i = 0; i < marketParamsList.length; i++) {
-                if (Id.unwrap(marketParamsList[i].id()) == Id.unwrap(marketId)) {
-                    marketParamsList[i] = marketParamsList[marketParamsList.length - 1];
-                    marketParamsList.pop();
-                    break;
-                }
-            }
-        } else if (oldAllocation == 0 && newAllocation > 0) {
-            marketParamsList.push(marketParams);
-        }
+        return (ids(), change);
     }
 
-    function allocation(MarketParams memory marketParams) public view returns (uint256) {
-        return IVaultV2(parentVault).allocation(keccak256(abi.encode("this/marketParams", address(this), marketParams)));
-    }
+    /* VIEWS */
 
     /// @dev Returns adapter's ids.
-    function ids(MarketParams memory marketParams) public view returns (bytes32[] memory) {
+    function ids() public view returns (bytes32[] memory) {
         bytes32[] memory ids_ = new bytes32[](3);
-        ids_[0] = adapterId;
-        ids_[1] = keccak256(abi.encode("collateralToken", marketParams.collateralToken));
-        ids_[2] = keccak256(abi.encode("this/marketParams", address(this), marketParams));
+        ids_[0] = morphoMarketV1Id;
+        ids_[1] = adapterId;
+        ids_[2] = collateralTokenId;
```

Comment:
We could put the adapter id as the last id. It would be consistent with vault v1 adapter, but more importantly it would make the id list go from the most general to the less general